### PR TITLE
Handle OperationCancelledExceptions to prevent log noise

### PR DESCRIPTION
**Pull request by Tratcher**
#150 @halter73 @muratg @benaadams 

Note the RequestAborted token may have been substituted out by middleware, we can't assume the socket is dead, so we must make sure that the response never completes with a 200 in this scenario.

### CODE REVIEW

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs [hunk 1]
```diff
@@ -346,23 +346,39 @@ public Task SendStatusAsync(int statusCode)
...
+                {
+                    readStream.Dispose();
+                }
             }
-            finally
+            catch (OperationCanceledException ex)
             {
-                readStream.Dispose();
+                _logger.LogWriteCancelled(ex);
+                if (_context.Response.HasStarted)
```
Comment (halter73) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs]: Couldn't `HasStarted` theoretically be replaced just like `RequestAborted`? I think trusting the `RequestAborted` token is reasonable enough. If you don't trust it, I would call `_context.Abort()` in any case.

**Comment by halter73** [comment]
Any tests?

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs [hunk 2]
```diff
@@ -346,23 +346,33 @@ public Task SendStatusAsync(int statusCode)
...
-                await StreamCopyOperation.CopyToAsync(readStream, _response.Body, _length, StreamCopyBufferSize, _context.RequestAborted);
+                string physicalPath = _fileInfo.PhysicalPath;
+                var sendFile = _context.Features.Get<IHttpSendFileFeature>();
+                if (sendFile != null && !string.IsNullOrEmpty(physicalPath))
+                {
+                    await sendFile.SendFileAsync(physicalPath, 0, _length, _context.RequestAborted);
+                    return;
+                }
+
+                Stream readStream = _fileInfo.CreateReadStream();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs]: Use a using statement?

File: src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs [hunk 3]
```diff
@@ -346,23 +346,33 @@ public Task SendStatusAsync(int statusCode)
...
-            Stream readStream = _fileInfo.CreateReadStream();
             try
             {
-                // Larger StreamCopyBufferSize is required because in case of FileStream readStream isn't going to be buffering
-                await StreamCopyOperation.CopyToAsync(readStream, _response.Body, _length, StreamCopyBufferSize, _context.RequestAborted);
+                string physicalPath = _fileInfo.PhysicalPath;
+                var sendFile = _context.Features.Get<IHttpSendFileFeature>();
+                if (sendFile != null && !string.IsNullOrEmpty(physicalPath))
+                {
+                    await sendFile.SendFileAsync(physicalPath, 0, _length, _context.RequestAborted);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs]: It interesting we care about passing in the client disconnected token when rendering static files in the first place. Couldn't we just ignore this by default and allow people to turn it on via a setting or something

Comment (Tratcher) [inline src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs]: hmm, maybe we don't need it for SendFileAsync, but we do want it for the manual copy below. Files may be large and we don't want to keep reading from disk if the writes are failing.

**Comment by Tratcher** [comment]
@halter73 Added tests

File: test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/Microsoft.AspNetCore.StaticFiles.FunctionalTests.xproj [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    <BaseIntermediateOutputPath Condition="'$(BaseIntermediateOutputPath)'=='' ">.\obj</BaseIntermediateOutputPath>
+    <OutputPath Condition="'$(OutputPath)'=='' ">.\bin\</OutputPath>
+  </PropertyGroup>
+  <PropertyGroup>
+    <SchemaVersion>2.0</SchemaVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
+  </ItemGroup>
+  <Import Project="$(VSToolsPath)\DNX\Microsoft.DNX.targets" Condition="'$(VSToolsPath)' != ''" />
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/Microsoft.AspNetCore.StaticFiles.FunctionalTests.xproj]: dnx? what

Comment (Tratcher) [inline test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/Microsoft.AspNetCore.StaticFiles.FunctionalTests.xproj]: @BrennanConroy that's still right isn't it? I copied this from the other test project

Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/Microsoft.AspNetCore.StaticFiles.FunctionalTests.xproj]: Yeah, thats fine. @dougbu has more details

**Comment by halter73** [comment]
Socket.Connect is throwing a PlatformNotSupportedException on Mac and Linux. [Travis build](https://travis-ci.org/aspnet/StaticFiles/builds/156611327)

    fail: Microsoft.AspNetCore.StaticFiles.StaticFileMiddlewareTests.ClientDisconnect_Kestrel_NoWriteExceptionThrown [FAIL]

    info: System.PlatformNotSupportedException : This platform does not support connecting sockets
    to DNS endpoints via the instance Connect and ConnectAsync methods, due to the potential for a host
    name to map to multiple IP addresses and sockets becoming invalid for use after a failed connect
    attempt. Use the static ConnectAsync method, or provide to the instance methods the specific
    IPAddress desired.

    info:       Stack Trace:
    info:            at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port)
    info:            at System.Net.Sockets.Socket.Connect(String host, Int32 port)
    info:         /home/travis/build/aspnet/StaticFiles/test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/StaticFileMiddlewareTests.cs(243,0): at Microsoft.AspNetCore.StaticFiles.StaticFileMiddlewareTests.SendSocketRequestAsync(String address, String path, String method)
    info:         /home/travis/build/aspnet/StaticFiles/test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/StaticFileMiddlewareTests.cs(219,0): at Microsoft.AspNetCore.StaticFiles.StaticFileMiddlewareTests.ClientDisconnect_NoWriteExceptionThrown(ServerType serverType)
    info:         /home/travis/build/aspnet/StaticFiles/test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/StaticFileMiddlewareTests.cs(167,0): at Microsoft.AspNetCore.StaticFiles.StaticFileMiddlewareTests.ClientDisconnect_Kestrel_NoWriteExceptionThrown()

File: test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/Microsoft.AspNetCore.StaticFiles.FunctionalTests.xproj (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.StaticFiles.FunctionalTests/Microsoft.AspNetCore.StaticFiles.FunctionalTests.xproj]: These files still exist and automatically handle both regular and web projects. This line also matches the template I use when rewriting large numbers of `.xproj` files.

**Comment by halter73** [comment]
:ship: Assuming travis passes.

**Pull request merged**